src/libcore/ptr.rs

```diff
@@ -228,6 +228,67 @@ pub const fn null<T>() -> *const T { 0 as *const T }
 #[rustc_promotable]
 pub const fn null_mut<T>() -> *mut T { 0 as *mut T }
 
+#[repr(C)]
+union Repr<T> {
+    rust: *const [T],
+    rust_mut: *mut [T],
+    raw: FatPtr<T>,
+}
+
+#[repr(C)]
+struct FatPtr<T> {
+    data: *const T,
+    len: usize,
+}
+
+/// Forms a slice from a pointer and a length.
+///
+/// The `len` argument is the number of **elements**, not the number of bytes.
+///
+/// # Panics
+///
+/// The total size of the slice should be no
+/// larger than `isize::MAX` **bytes** in memory.
```

Comment:
This panic happens only when debug assertions are enabled (so basically never, because the libstd we distribute is built without debug assertions).

Without that check, I am not sure if this method is safe. @rust-lang/wg-unsafe-code-guidelines do you think the validity invariant for raw slices might require that the slice is not too big? I feel like probably, the validity invariant for fat raw pointers does not impose any requirement, but that would mean that calling `size_of_val` on a raw pointer can cause UB.

Reply:
I agree that there shouldn't be any validity invariant for raw slices. `size_of_val` takes a reference anyway, so it has plenty of ways to cause UB even if the total length of the slice is less than `isize::MAX` bytes. Plus, if we ever gain a raw pointer equivalent to that (e.g., as part of custom DSTs) then I think that one can be defined to be an integer multiply without UB.

Reply:
Something I noticed in some experiments I did with unsized type allocation is that we might want a validity invariant for raw slices to be able to always (safely) determine the layout of the memory that should be used for the slice; if slice fat pointers can be safely created where the length exceeds `usize::MAX` API's around that become more difficult to design.

Not sure I know what is or isn't the correct decision there, but it is a decision to be made.

Reply:
It's already possible to create such oversized raw slices in safe Rust today by doing things like:

```rust
123456789 as *const [(); usize::max_value()] as *const [()] as *const [u64]
```

So these functions aren't really letting you do anything new.

Reply:
Good catch! We should probably remove that part from the docs then (and add a comment about why we allow just about any value in raw slices).